### PR TITLE
ONNX: Support renamed input_names in dynamic_shapes for torch.onnx.export

### LINKED ISSUE #166001: `torch.onnx.export`'s `dynamic_shape` should take renamed inputs into account
### 🐛 Describe the bug

Presently, `torch.onnx.export` (with `dynamo=True`) does not take into account renamed inputs and errors out/complains if the explicit labels assigned to the inputs is used in place of the default (e.g. `x` or `x1`) names.

IMHO, the correct behavior here should be that input names are remapped prior to `dynamic_shape` being used/verified, and the renamed inputs should be the only names acceptable to the `dynamic_shape` dict keys. If you wish to _additionally_ support the un-renamed input variables as dict keys, perhaps that's understandable, but the renamed inputs should definitely be supported.

Example:

```python
    onnx_path = "foo.onnx"
    onnx_program = torch.onnx.export(
        model,
        (bgr_tensor_input,),
        input_names=["bgr_image"],
        output_names=["embeddings", "norm"],
        dynamic_shapes={
            "bgr_image": {0: "batch"},
        },
        dynamo=True,
    )
    onnx_program.save(onnx_path)
```

throws the following `torch.onnx._internal.exporter._errors.TorchExportError` exception:

> `<class 'torch._dynamo.exc.UserError'>`: When `dynamic_shapes` is specified as a dict, its top-level keys must be the arg names `['x']` of `inputs`, but here they are `['bgr_image']`. Alternatively, you could also ignore arg names entirely and specify `dynamic_shapes` as a list/tuple matching `inputs`. For more information about this error, see: https://pytorch.org/docs/main/generated/exportdb/index.html#dynamic-shapes-validation

### Versions

```
PyTorch version: 2.9.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 IoT Enterprise LTSC (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: 21.1.2
CMake version: version 3.31.6-msvc6
Libc version: N/A

Python version: 3.13.7 (tags/v3.13.7:bcee1c3, Aug 14 2025, 14:15:11) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26100-SP0
Is CUDA available: False
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080
Nvidia driver version: 581.29
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Name: AMD Ryzen Threadripper 1950X 16-Core Processor
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3600
MaxClockSpeed: 3600
L2CacheSize: 8192
L2CacheSpeed: None
Revision: 257

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

cc @justinchuby @titaiwangms @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### PR DESCRIPTION
**Pull request by sam-shubham**
## Summary

When `torch.onnx.export(dynamo=True)` is used with both `input_names`
(to rename ONNX inputs) and `dynamic_shapes` (dict form keyed by those
renamed names), export fails validation because `torch.export.export`
expects `dynamic_shapes` keys to match the original `forward()` parameter
names (e.g., `"x"`), not the renamed ONNX input names (e.g., `"bgr_image"`).

This PR remaps `dynamic_shapes` keys from `input_names` back to the
original model parameter names before calling `torch.export.export`.

---

## Root Cause

`dynamic_shapes` was previously passed directly to `torch.export.export`
without accounting for `input_names`. Since ONNX renaming occurs later via
`_ir_passes.rename_inputs`, validation would fail early when dict keys did
not match the original `forward()` signature.

---

## Fix

Adds `remap_dynamic_shapes_from_input_names` in
`torch/onnx/_internal/exporter/_dynamic_shapes.py` and invokes it from
`export_compat()` before calling `torch.export.export`.

---

## Example
Previously Failing, Now Works

```python
torch.onnx.export(
    model,
    (input_tensor,),
    "model.onnx",
    input_names=["bgr_image"],
    dynamic_shapes={"bgr_image": {0: "batch"}},
    dynamo=True,
)

### CODE REVIEW

File: test/onnx/exporter/test_dynamic_shapes.py [hunk 1]
```diff
@@ -629,6 +629,179 @@ def test__any_str_or_dim_in_dynamic_shapes_returns_true(self):
...
+        # dynamic_shapes uses both renamed names, which after remapping
+        # would both resolve correctly without collision.
+        dynamic_shapes = {"y": {0: "batch"}, "x": {0: "batch"}}
+        input_names = ["y", "x"]
+        result = _dynamic_shapes.remap_dynamic_shapes_from_input_names(
+            model, dynamic_shapes, input_names
+        )
+        # "y" (input_name[0]) → param "x", "x" (input_name[1]) → param "y"
+        self.assertEqual(result, {"x": {0: "batch"}, "y": {0: "batch"}})
+
```
Comment (titaiwangms) [inline test/onnx/exporter/test_dynamic_shapes.py:804]: Can you add three more tests: 
(1) nested inputs: torch.export prefers nested structure, while onnx model accepts flattened input_names, so we need to make sure the implementation working on nested structures: https://github.com/pytorch/pytorch/blob/54fbebf42930215f66af5ad6cadb8d99391f4a6e/test/onnx/exporter/test_dynamic_shapes.py#L168
(2) LLM is the most popular architecture now, so please demonstrate the usage in here: https://github.com/pytorch/pytorch/blob/main/test/onnx/exporter/test_small_models_e2e.py
(3) A dummy nested input case in https://github.com/pytorch/pytorch/blob/2c0c5cba4babec967264c832210c916fdcba1c6c/test/onnx/exporter/test_small_models_e2e.py#L528

File: torch/onnx/_internal/exporter/_dynamic_shapes.py [hunk 2]
```diff
@@ -336,6 +336,88 @@ def is_axes(x) -> bool:
...
+
+    # Build a mapping: input_name -> original_param_name
+    # input_names may be longer than the number of forward parameters (for example,
+    # when additional names are provided that do not correspond to inputs), so only
+    # map up to the number of model parameters. This remapping assumes a 1:1,
+    # ordered correspondence between input_names and the model's forward parameters.
+    input_name_to_param: dict[str, str] = {}
+    for i, param_name in enumerate(original_param_names):
+        if i < len(input_names):
+            input_name_to_param[input_names[i]] = param_name
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_dynamic_shapes.py:465]: This does not look like handling nested input cases.

File: test/onnx/exporter/test_small_models_e2e.py [hunk 3]
```diff
@@ -594,6 +594,110 @@ def forward(
...
+                return x + y, c
+
+        input = (
+            torch.ones(5),
+            [torch.zeros(5), torch.ones(5)],
+            {"a": torch.zeros(5), "b": torch.ones(5)},
+            torch.ones(6),
+        )
+        input_names = ["input_x", "input_y0", "input_y1", "input_a", "input_b", "input_c"]
+        dynamic_shapes = {
```
Comment (titaiwangms) [inline test/onnx/exporter/test_small_models_e2e.py:655]: `dynamic_shapes` does not follow onnx model input format. It follows nn.Module which in this case, should be nested as input format. https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/export/api_reference.html#torch.export.export

It's defined by torch.export.export.

Comment (sam-shubham) [reply test/onnx/exporter/test_small_models_e2e.py:655]: ohh sorry for that!! fixed it in the latest commit...

File: test/onnx/exporter/test_small_models_e2e.py [hunk 4]
```diff
@@ -594,6 +594,117 @@ def forward(
...
+            "input_c",
+        ]
+        # dynamic_shapes follows nn.Module parameter structure, not flat input_names
+        dim = torch.export.Dim("dim", min=3)
+        dim_c = torch.export.Dim("dim_c", min=3)
+        dynamic_shapes = {
+            "x": {0: dim},
+            "ys": [{0: dim}, {0: dim}],
+            "zs": {"a": {0: dim}, "b": {0: dim}},
+            "c": {0: dim_c},
```
Comment (titaiwangms) [inline test/onnx/exporter/test_small_models_e2e.py]: This is not using input_names

File: test/onnx/exporter/test_small_models_e2e.py [hunk 5]
```diff
@@ -594,6 +594,117 @@ def forward(
...
+        onnx_program = torch.onnx.export(
+            model,
+            (input_ids, attention_mask),
+            dynamo=True,
+            verbose=False,
+            input_names=["tokens", "mask"],
+            output_names=["logits"],
+            dynamic_shapes={
+                "tokens": {0: batch_dim, 1: seq_dim},
+                "mask": {0: batch_dim, 1: seq_dim},
```
Comment (titaiwangms) [inline test/onnx/exporter/test_small_models_e2e.py]: In terms of LLM, we are looking for a case with kv cache (nested) involved.

**Review by titaiwangms — CHANGES_REQUESTED** [review]

File: test/onnx/exporter/test_small_models_e2e.py (hunk 4 above)
Comment (sam-shubham) [reply test/onnx/exporter/test_small_models_e2e.py]: Updated!!
Used Flat `input_names`-Keyed `dynamic_shapes`

So it actually exercises the remapping logic:

```python
input_names = ["input_x", "input_y0", "input_y1", "input_a", "input_b", "input_c"]

dynamic_shapes = {
    "input_x": {0: dim},
    "input_y0": {0: dim},
    "input_y1": {0: dim},
    "input_a": {0: dim},
    "input_b": {0: dim},
    "input_c": {0: dim_c},
}
```

File: test/onnx/exporter/test_small_models_e2e.py (hunk 5 above)
Comment (sam-shubham) [reply test/onnx/exporter/test_small_models_e2e.py]: Updated!!
Replaced with a Model that Has `past_key_values` as Nested Input

```python
class LLMWithKVCache(torch.nn.Module):
    def __init__(self, vocab_size=1000, hidden_dim=64):
        super().__init__()
        self.embed = torch.nn.Embedding(vocab_size, hidden_dim)
        self.linear = torch.nn.Linear(hidden_dim, hidden_dim)

    def forward(self, input_ids, attention_mask, past_key_values: list[torch.Tensor]):
        x = self.embed(input_ids)
        x = x * attention_mask.unsqueeze(-1)
        for kv in past_key_values:
            x = x + kv
        return self.linear(x)
```

```python
input_names = ["tokens", "mask", "past_kv_0", "past_kv_1"]

dynamic_shapes = {
    "tokens": {0: batch_dim, 1: seq_dim},
    "mask": {0: batch_dim, 1: seq_dim},
    "past_kv_0": {0: batch_dim, 1: seq_dim},
    "past_kv_1": {0: batch_dim, 1: seq_dim},
}
```
